pkg/common/containerinfo/extract.go

```diff
@@ -0,0 +1,280 @@
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/spiffe/spire/pkg/agent/common/cgroups"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/mount-utils"
+)
+
+var (
+	// This regex covers both the cgroupfs and systemd rendering of the pod
+	// UID. The dashes are replaced with underscores in the systemd rendition.
+	rePodUID = regexp.MustCompile(`\b(?:pod([[:xdigit:]]{8}[-_][[:xdigit:]]{4}[-_][[:xdigit:]]{4}[-_][[:xdigit:]]{4}[-_][[:xdigit:]]{12}))\b`)
+
+	// The container ID is a 64-character hex string, by convention.
+	reContainerID = regexp.MustCompile(`\b([[:xdigit:]]{64})\b`)
+
+	// underToDash replaces underscores with dashes. The systemd cgroups driver
+	// doesn't allow dashes so the pod UID component has dashes replaced with
+	// underscores by the Kubelet.
+	underToDash = strings.NewReplacer("_", "-")
+)
+
+type Extractor struct {
+	RootDir string
+}
+
+func (e *Extractor) GetContainerID(pid int, log hclog.Logger) (string, error) {
+	_, containerID, err := e.extractInfo(pid, log, false)
+	return containerID, err
+}
+
+func (e *Extractor) GetPodUIDAndContainerID(pid int, log hclog.Logger) (types.UID, string, error) {
+	return e.extractInfo(pid, log, true)
+}
+
+func (e *Extractor) extractInfo(pid int, log hclog.Logger, extractPodUID bool) (types.UID, string, error) {
+	// Try to get the information from /proc/pid/mountinfo first. Otherwise,
+	// fall back to /proc/pid/cgroup. If it isn't in mountinfo, then the
+	// workload being attested likely originates in the same Pod as the agent.
+	//
+	// It may not be possible to attest a process running in the same container
+	// as the agent because, depending on how cgroups are being used,
+	// /proc/<pid>/mountinfo or /proc/<pid>/cgroup may not contain any
+	// information on the container ID or pod.
+
+	podUID, containerID, err := e.extractPodUIDAndContainerIDFromMountInfo(pid, log, extractPodUID)
+	if err != nil {
+		return "", "", err
+	}
+
+	if containerID == "" {
+		podUID, containerID, err = e.extractPodUIDAndContainerIDFromCGroups(pid, log, extractPodUID)
+		if err != nil {
+			return "", "", err
+		}
+	}
+
+	return podUID, containerID, nil
+}
+
+func (e *Extractor) extractPodUIDAndContainerIDFromMountInfo(pid int, log hclog.Logger, extractPodUID bool) (types.UID, string, error) {
+	mountInfoPath := filepath.Join(e.RootDir, "/proc", fmt.Sprint(pid), "mountinfo")
+
+	mountInfos, err := mount.ParseMountInfo(mountInfoPath)
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return "", "", nil
+		}
+		return "", "", status.Errorf(codes.Internal, "failed to parse mount info at %q: %v", mountInfoPath, err)
+	}
+
+	for i, mountInfo := range mountInfos {
+		log.Debug("PID mount enumerated",
+			"index", i+1,
+			"total", len(mountInfos),
+			"type", mountInfo.FsType,
+			"root", mountInfo.Root,
+		)
+	}
```

Comment:
It could be pretty noisy. This information is pretty important when trying to debug a problem though. Maybe we can add an extra configurable to the plugins.... something like "log_verbose_container_locator_debug_info"? And only log these lines if that is set?

Reply:
I like that idea, basically allow only if you want to

Reply:
added config